lighthouse-core/audits/byte-efficiency/render-blocking-resources.js

```diff
@@ -0,0 +1,213 @@
+      if (wastedMs < MINIMUM_WASTED_MS) continue;
+
+      results.push({
+        url: resource.tag.url,
+        totalBytes: resource.transferSize,
+        wastedMs,
+      });
+    }
+
+    if (!results.length) {
+      return {results, wastedMs: 0};
+    }
+
+    const wastedMs = RenderBlockingResources.estimateSavingsWithGraphs(
+      simulator,
+      fcpSimulation.optimisticGraph,
+      deferredNodeIds,
+      wastedBytesMap
+    );
+
+    return {results, wastedMs};
+  }
+
+  /**
+   * Estimates how much faster this page would reach FCP if we inlined all the used CSS from the
+   * render blocking stylesheets and deferred all the scripts. This is more conservative than
+   * removing all the assets and more aggressive than inlining everything.
+   *
+   * *Most* of the time, scripts in the head are there accidentally/due to lack of awareness
+   * rather than necessity, so we're comfortable with this balance. In the worst case, we're telling
+   * devs that they should be able to get to a reasonable first paint without JS, which is not a bad
+   * thing.
+   *
+   * @param {Simulator} simulator
+   * @param {Node} fcpGraph
+   * @param {Set<string>} deferredIds
+   * @param {Map<string, number>} wastedBytesMap
+   * @return {number}
+   */
+  static estimateSavingsWithGraphs(simulator, fcpGraph, deferredIds, wastedBytesMap) {
+    const originalEstimate = simulator.simulate(fcpGraph).timeInMs;
+
+    let totalChildNetworkBytes = 0;
+    const minimalFCPGraph = fcpGraph.cloneWithRelationships(node => {
+      const canDeferRequest = deferredIds.has(node.id);
+      const isStylesheet =
+        node.type === Node.TYPES.NETWORK &&
+        node.record._resourceType === WebInspector.resourceTypes.Stylesheet;
+      if (canDeferRequest && isStylesheet) {
+        // We'll inline the used bytes of the stylesheet and assume the rest can be deferred
+        const wastedBytes = wastedBytesMap.get(node.record.url) || 0;
+        totalChildNetworkBytes += node.record._transferSize - wastedBytes;
+      }
+
+      // If a node can be deferred, exclude it from the new FCP graph
+      return !canDeferRequest;
+    });
+
+    // Add the inlined bytes to the HTML response
+    minimalFCPGraph.record._transferSize += totalChildNetworkBytes;
+    const estimateAfterInline = simulator.simulate(minimalFCPGraph).timeInMs;
+    minimalFCPGraph.record._transferSize -= totalChildNetworkBytes;
+    return Math.round(Math.max(originalEstimate - estimateAfterInline, 0));
+  }
+
+  /**
+   * @param {!Artifacts} artifacts
+   * @param {LH.Audit.Context} context
+   * @return {Map<string, number>}
+   */
+  static async computeWastedCSSBytes(artifacts, context) {
+    const wastedBytesByUrl = new Map();
+    try {
+      const results = await UnusedCSS.audit(artifacts, context);
+      for (const item of results.details.items) {
+        wastedBytesByUrl.set(item.url, item.wastedBytes);
+      }
+    } catch (_) {}
```

Comment:
whats the failure mode here?

Reply:
assume all bytes were used and inline all CSS